### PR TITLE
[OSDOCS-2604] Document ShiftStack cloud provider option configuration

### PR DESCRIPTION
**Pull request by maxwelldb**
Context: 

- https://issues.redhat.com/browse/OSDOCS-2604
- https://issues.redhat.com/browse/OSASINFRA-2412

Example: https://deploy-preview-36088--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_openstack/installing-openstack-installer-custom.html#installation-osp-setting-cloud-provider-options_installing-openstack-installer-custom

4.9.

### CODE REVIEW

**Comment by maxwelldb** [comment]
@mandre Is this essentially what's required for OSASINFRA-2412, at least as far as configuration at install-time goes? 

https://deploy-preview-36088--osdocs.netlify.app/openshift-enterprise/latest/installing/installing_openstack/installing-openstack-installer-custom.html#installation-osp-setting-cloud-provider-options_installing-openstack-installer-custom

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+[source,yaml]
+----
+#...
+[LoadBalancer]
+use-octavia=true <1>
+lb-provider = "amphora" <2>
+floating-network-id="d3deb660-4190-40a3-91f1-37326fe6ec4a"<3>
+floating-subnet-id="4a2fe995-84d4-4eac-8bcf-c5624f63dba0"
+subnet-id="e3d26bd3-0b6b-4c4d-b184-71c5be59b8b2"
+create-monitor=true
```
Comment (mandre) [inline modules/installation-osp-setting-cloud-provider-options.adoc:42]: The `create-monitor=true` option was not tested by QE as far as I can tell, I'd suggest removing it from the example until we confirmed it works as expected.

**Review by mandre — COMMENTED** [review]
Looks really nice. Adding Michal and Emilio who worked on the epic.
/cc @dulek @iamemilio

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+[TIP]
+====
+You can update your cloud provider configuration after you run the installer. On a command line, run:
+
+[source,terminal]
+----
+$ oc edit configmap -n openshift-config cloud-provider-config
+----
+
+After you save your changes, your cluster will take some time to reconfigure itself. The process is complete if none of your nodes have a `SchedulingDisabled` status.
```
Comment (dulek) [inline modules/installation-osp-setting-cloud-provider-options.adoc:63]: I'd add a warning that it is important to double-check the file structure before saving it. We've seen clusters breaking if e.g. options were put in wrong sections.

File: modules/installation-osp-setting-cloud-provider-options.adoc (hunk 1 above)
Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc:42]:
```suggestion
```

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+monitor-timeout=30s
+monitor-max-retries=5
+#...
+----
+<1> This property enables Octavia integration.
+<2> This property sets the Octavia provider that your load balancer uses. It accepts `"ovn"` or `"amphora`" as values. If you choose to use OVN, you must also set `lb-method` to `SOURCE_IP_PORT`.
+<3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.
+
+. Save the changes to the file and proceed with installation.
+
```
Comment (maxwelldb) [inline modules/installation-osp-setting-cloud-provider-options.adoc:53]: FYI @dulek from https://github.com/openshift/openshift-docs/pull/36088/files/80343fcc4cceca59c81751af820d6ca0979062d8#r707485144

```suggestion

[IMPORTANT]
====
Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.
====

. Save the changes to the file and proceed with installation.

```

Comment (dulek) [reply modules/installation-osp-setting-cloud-provider-options.adoc:52]: Looks great!

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+#...
+[LoadBalancer]
+use-octavia=true <1>
+lb-provider = "amphora" <2>
+floating-network-id="d3deb660-4190-40a3-91f1-37326fe6ec4a"<3>
+floating-subnet-id="4a2fe995-84d4-4eac-8bcf-c5624f63dba0"
+subnet-id="e3d26bd3-0b6b-4c4d-b184-71c5be59b8b2"
+monitor-delay=60s
+monitor-timeout=30s
+monitor-max-retries=5
```
Comment (mandre) [inline modules/installation-osp-setting-cloud-provider-options.adoc:44]: You should also remove these options that no longer make sense without `create-monitor=true`

Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc:44]:
```suggestion
```

**Comment by maxwelldb** [comment]
@mandre @dulek @iamemilio Any objections to me moving this to QE?

**Comment by maxwelldb** [comment]
@eurijon Any updates on this?

**Review by mandre — COMMENTED** [review]
LGTM

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
++
+[source,yaml]
+----
+#...
+[LoadBalancer]
+use-octavia=true <1>
+lb-provider = "amphora" <2>
+floating-network-id="d3deb660-4190-40a3-91f1-37326fe6ec4a"<3>
+floating-subnet-id="4a2fe995-84d4-4eac-8bcf-c5624f63dba0"
+subnet-id="e3d26bd3-0b6b-4c4d-b184-71c5be59b8b2"
```
Comment (dulek) [inline modules/installation-osp-setting-cloud-provider-options.adoc:41]: Hm, we should probably either explain this or just omit this option (cloud-provider will autodetect it). Basically it's about in which subnet to create the load balancers. We've never tested anything other than the subnet used by the nodes, so I'd rather remove this line.

@iamemilio: I just realized that we haven't checked how the autodetection works if you have nodes in another subnet. There may be dragons.

**Comment by dulek** [comment]
Minor comment inline, besides that it looks okay to me.

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+<1> This property enables Octavia integration.
+<2> This property sets the Octavia provider that your load balancer uses. It accepts `"ovn"` or `"amphora`" as values. If you choose to use OVN, you must also set `lb-method` to `SOURCE_IP_PORT`.
+<3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.
+
+[IMPORTANT]
+====
+Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.
+====
+
+. Save the changes to the file and proceed with installation.
```
Comment (eurijon) [inline modules/installation-osp-setting-cloud-provider-options.adoc:53]: In the preview this sentence is preceded by the index n.1 when it should n.3 I think

**Review by eurijon — COMMENTED** [review]
Minor format comment, besides that it lgtm.

File: modules/installation-osp-setting-cloud-provider-options.adoc (hunk 5 above)
Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc:41]: @eurijon No qualms with removing this line? 

```suggestion
```

File: modules/installation-osp-setting-cloud-provider-options.adoc (hunk 6 above)
Comment (maxwelldb) [inline modules/installation-osp-setting-cloud-provider-options.adoc:53]:
```suggestion
<1> This property enables Octavia integration.
<2> This property sets the Octavia provider that your load balancer uses. It accepts `"ovn"` or `"amphora`" as values. If you choose to use OVN, you must also set `lb-method` to `SOURCE_IP_PORT`.
<3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.
+
[IMPORTANT]
====
Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.
====

. Save the changes to the file and proceed with installation.
```

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+<3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.
++
+[IMPORTANT]
+====
+Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.
+====
+
+. Save the changes to the file and proceed with installation.
+
+[TIP]
```
Comment (bobfuru) [inline modules/installation-osp-setting-cloud-provider-options.adoc:55]: This should get a preceding `+` for indentation.

**Review by bobfuru — COMMENTED** [review]
A few minor nits and otherwise LGTM!

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+= Setting cloud provider options
+
+Optionally, you can edit the cloud provider configuration for your cluster. The cloud provider configuration controls how {product-title} interacts with {rh-openstack-first}.
+
+For a complete list of cloud provider configuration parameters, see link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].
+
+.Prerequisites
+
+* Have manifest files generated for your cluster.
+** If you don't have manifest files, generate them by running `openshift-install --dir <destination_directory> create manifests` from a command line.
```
[the remark this answers is not in the document]
Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc:18]:
```suggestion
** If you do not have manifest files, generate them by running `openshift-install --dir <destination_directory> create manifests` from a command line.
```

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+subnet-id="e3d26bd3-0b6b-4c4d-b184-71c5be59b8b2"
+#...
+----
+<1> This property enables Octavia integration.
+<2> This property sets the Octavia provider that your load balancer uses. It accepts `"ovn"` or `"amphora`" as values. If you choose to use OVN, you must also set `lb-method` to `SOURCE_IP_PORT`.
+<3> This property is required if you want to use multiple external networks with your cluster. The cloud provider creates floating IP addresses on the network that is specified here.
++
+[IMPORTANT]
+====
+Prior to saving your changes, verify that the file is structured correctly. Clusters may fail if properties are not placed in the appropriate section.
```
[the remark this answers is not in the document]
Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc:50]:
```suggestion
Prior to saving your changes, verify that the file is structured correctly. Clusters might fail if properties are not placed in the appropriate section.
```

File: modules/installation-osp-setting-cloud-provider-options.adoc (hunk 7 above)
Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc:57]: Done in 57cb23e87ad98ea28dab195fd03a43869c8f9d9f

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+[id="installation-osp-setting-cloud-provider-options_{context}"]
+= Setting cloud provider options
+
+Optionally, you can edit the cloud provider configuration for your cluster. The cloud provider configuration controls how {product-title} interacts with {rh-openstack-first}.
+
+For a complete list of cloud provider configuration parameters, see link:https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf[the `cloud.conf` specification].
+
+.Prerequisites
+
+* Have manifest files generated for your cluster.
```
[the remark this answers is not in the document]
Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc]: Done in 07fa501914a5247a531e42706b2c430fe3a41262

File: modules/installation-osp-setting-cloud-provider-options.adoc [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+Configuring Octavia for load balancing is a common case. For example:
++
+[source,yaml]
+----
+#...
+[LoadBalancer]
+use-octavia=true <1>
+lb-provider = "amphora" <2>
+floating-network-id="d3deb660-4190-40a3-91f1-37326fe6ec4a"<3>
+floating-subnet-id="4a2fe995-84d4-4eac-8bcf-c5624f63dba0"
```
Comment (mandre) [inline modules/installation-osp-setting-cloud-provider-options.adoc:42]: `floating-subnet-id` is not a valid option according to https://v1-18.docs.kubernetes.io/docs/concepts/cluster-administration/cloud-providers/#cloud-conf.

Setting it causes errors in openshift-kube-controller-manager:
```
F0929 11:19:55.041313       1 controllermanager.go:246] error building controller context: cloud provider could not be initialized: could not init cloud provider "openstack": warning:                                                      
can't store data at section "LoadBalancer", variable "floating-subnet-id"
```

**Review by mandre — CHANGES_REQUESTED** [review]

File: modules/installation-osp-setting-cloud-provider-options.adoc (hunk 11 above)
Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc:42]: Gone!
```suggestion
```

File: modules/installation-osp-setting-cloud-provider-options.adoc (hunk 5 above)
Comment (eurijon) [reply modules/installation-osp-setting-cloud-provider-options.adoc:42]: It can be removed as it's autodetected by the cloud-provider

Comment (maxwelldb) [reply modules/installation-osp-setting-cloud-provider-options.adoc:42]: TY.

**Comment by eurijon** [comment]
It LGTM

**Pull request merged**